python/pyarrow/dataset.py

```diff
@@ -252,7 +243,18 @@ def factory(path_or_paths, filesystem=None, partitioning=None,
     elif isinstance(partitioning, Partitioning):
         options.partitioning = partitioning
 
-    return FileSystemDatasetFactory(fs, paths_or_selector, format, options)
+    factories = []
+    for path in path_or_paths:
+        fs, paths_or_selector = _ensure_fs_and_paths(path, filesystem)
+        factories.append(FileSystemDatasetFactory(fs, paths_or_selector,
```

Comment:
Slight downside of this is that if you now create a dataset from a list of file paths, you get a UnionDataset instead of a FileSystemDataset. 

We might need to think about what the preferred API is to create a dataset this way. 
Doing `ds.FileSystemDataset(..)` is bit hard / low-level. But needing to do `ds.dataset(ds.factory([..]))` as before was also not that nice.